### ISSUE TITLE
Go parts of `quicktest.sh` and `lit` tests fail because of lack of `GO111MODULE=off`

### ISSUE THREAD

**Issue by hmijail** [dafny-lang/dafny#1156]
macOS 11.2.3
Dafny 3.0.0-30204 (built from source at commit d25ae2577b5d38b3b4df90a7e8062d3f6fbfd2a7)
golang 1.16.2

quicktests.sh ended in failure like this:
```
...
Compiling to Go

Dafny program verifier finished with 0 verified, 0 errors
c-go/src/c.go:6:3: package System_ is not in GOROOT (/usr/local/Cellar/go/1.16.2/libexec/src/System_)
c-go/src/c.go:5:3: package dafny is not in GOROOT (/usr/local/Cellar/go/1.16.2/libexec/src/dafny)
```
(which, by the way, as a go hobbyist and dafny noob, was hard for me to ascertain as failure or warning or what)

I could fix it with `export GO111MODULE=off`.

A number of `lit` tests were also failing, and the `export` wasn't helping. I could fix it with
```
go env -w GO111MODULE=off
```
... though IIUC this might interfere with other Go projects. I don't have any in flight so I don't mind.

Relatedly, in my 6-Core Core i7 Macbook, the lit tests run ~3x faster if I limit the workers with `lit -j 8 .`.

**Comment by hmijail** [comment]
For context: I am not interested in compiling Dafny to Go, I just wanted the tests to pass to make sure that my local build was reliable. If there was a way to identify the problem as a go thing and to disable the go parts I'd have used that.

**Comment by tchajed** [comment]
What's happening here is that Dafny emits code in the old Go project format with a `src` directory, expecting `$GOPATH` to be set to the output directory. Go has had module support for a while and in Go 1.16 modules became the default, so this old structure no longer compiles (see https://golang.org/ref/mod#mod-commands). You can restore the old behavior by setting `GO111MODULE=auto` (or `GO111MODULE=off` as @hmijail suggested), and I agree that this is the right solution for now.

Longer-term, Dafny should start supporting compiling Go to a module-aware format.

**Comment by RustanLeino** [comment]
Thanks for reporting the bug and the fix. I had tried to repro it too quickly before and didn't realize I was using an ancient version of Go.

**Comment by hmijail** [comment]
Revisiting this because I got bitten again, now that I'm looking into translating to Go. 

`GO111MODULE=auto` is the fix that got implemented and is mentioned in the docs. However, the meaning of `auto` changes depending on the Go version, and as of versions >1.17 it means `on`... with the exception of being inside of the GOPATH, where it means `off`.[(reference here)](https://dev.to/maelvls/why-is-go111module-everywhere-and-everything-about-go-modules-24k).

Fortunately, the docs _also_ mention changing the GOPATH, so it ends up working OK, but I'd say it's an obscure and fragile state which takes some digging to recreate inside of an IDE like VSCode.

So, until Go modules are supported, I'd suggest that `GO111MODULE` should be `off` in the code and in the docs.
